internal/controller/common.go

```diff
@@ -3,32 +3,258 @@ package controller
 	"fmt"
 
-	add "github.com/kyma-project/helm-broker/internal/addon"
-	"github.com/kyma-project/helm-broker/internal/controller/addons"
+	"time"
+
+	"github.com/Masterminds/semver"
+	"github.com/kyma-project/helm-broker/internal"
+	"github.com/kyma-project/helm-broker/internal/addon"
+	"github.com/kyma-project/helm-broker/internal/controller/repository"
+	"github.com/kyma-project/helm-broker/internal/storage"
 	"github.com/kyma-project/helm-broker/pkg/apis/addons/v1alpha1"
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/helm/pkg/proto/hapi/chart"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
-type addonLoader struct {
+type common struct {
 	addonGetterFactory addonGetterFactory
-	log                logrus.FieldLogger
-	dstPath            string
+
+	addonStorage addonStorage
+	chartStorage chartStorage
+
+	brokerSyncer brokerSyncer
+	brokerFacade brokerFacade
+
+	docsProvider docsProvider
+	protection   protection
+	dstPath      string
+
+	// used to distinguish namespace-scoped and cluster-wide addons configurations
+	namespace internal.Namespace
+
+	commonClient commonClient
+	log          logrus.FieldLogger
+
+	trace string
+}
+
+func newControllerCommon(client client.Client, addonGetterFactory addonGetterFactory, addonStorage addonStorage, chartStorage chartStorage, docsProvider docsProvider, brokerSyncer brokerSyncer, brokerFacade brokerFacade, dstPath string, log logrus.FieldLogger) *common {
+	return &common{
+		addonGetterFactory: addonGetterFactory,
+
+		addonStorage: addonStorage,
+		chartStorage: chartStorage,
+
+		brokerFacade: brokerFacade,
+		brokerSyncer: brokerSyncer,
+
+		docsProvider: docsProvider,
+		protection:   protection{},
+
+		namespace:    internal.ClusterWide,
+		commonClient: NewCommonClient(client, log),
+
+		dstPath: dstPath,
+		log:     log,
+	}
+}
+
+// SetWorkingNamespace sets services's working namespace. It should only be used by namespace-scoped controller.
+func (c *common) SetWorkingNamespace(namespace string) {
+	c.namespace = internal.Namespace(namespace)
+	for _, svc := range []NamespacedService{
+		c.brokerSyncer, c.brokerFacade, c.docsProvider, c.commonClient,
+	} {
+		svc.SetNamespace(namespace)
+	}
+}
+
+// Reconcile handles configuration state change
+func (c *common) Reconcile(addon *internal.CommonAddon, trace string) (reconcile.Result, error) {
+	if addon.Meta.DeletionTimestamp != nil {
+		c.log.Infof("Start delete %s", trace)
+
+		if err := c.OnDelete(addon); err != nil {
+			c.log.Errorf("while deleting %s process: %v", trace, err)
+			return reconcile.Result{RequeueAfter: time.Second * 15}, errors.Wrapf(err, "while deleting %s", trace)
+		}
+		c.log.Infof("Delete %s process completed", trace)
+		return reconcile.Result{}, nil
+	}
+
+	if addon.Status.ObservedGeneration == 0 {
+		c.log.Infof("Start add %s process", trace)
+
+		if err := c.PrepareForProcessing(addon); err != nil {
```

Comment:
Update process in the middle of the reconciliation causes each creation of new addons has two cycles after update finalizer/status `addons` falls into the next update cycle (maybe we should decrease `ObservedGeneration` in this step?)

Reply:
I managed to fix it by setting addon.Generation to the processed object after the update in the `common_client.go` 👍